tests/test_metrics.py

```diff
@@ -0,0 +1,174 @@
+    in a swarm, at a given point x, might tend to be better than any given one.
+    """
+    mean_point_preds = metrics.summarise_across(ypreds)
+    loss_mean_preds = metrics.rms_error(mean_point_preds, y)
+    assert loss_mean_preds.shape == (2,)
+
+
+def test_range_sd_point_pred():
+    """
+    This is as example of interest, since a large range in the sd of predictions will indicate the formation of
+    'knots' or 'nodes' in the prediction domain.  I.e. all the bees are making very similar predictions at some
+    point in the domain, and but are much more dispersed in others.
+    """
+    sd_point_preds = metrics.summarise_across(ypreds, "std")
+    range_sd_point_preds = metrics.summarise_across(sd_point_preds, "range", 'x')
+    assert range_sd_point_preds.shape == (2,)
+
+
+def test_median_sd_point_pred():
+    """
+    This is an example of interest, since it represents the degree to which the different bees are tending to
+    cohere across the whole domain, and produce similar predictions (regardless of whether the predictions are
+    very good or not).
+    """
+    sd_point_preds = metrics.summarise_across(ypreds, "std")
+    med_sd_point_preds = metrics.summarise_across(sd_point_preds, "median", 'x')
+    assert med_sd_point_preds.shape == (2,)
+
+
+def test_if_nom_first_below():
+    epoch = metrics.iteration_finder(epoch_scores, 0.25, "nominal", "first", "below")
+    assert epoch_scores[epoch] <= 0.25
+    assert epoch_scores[epoch - 1] > 0.25
+    assert metrics.iteration_finder(epoch_scores, 0.001, "nominal", "first", "below") is None
+
+
+def test_if_nom_always_below():
+    epoch = metrics.iteration_finder(epoch_scores, 0.25, "nominal", "always", "below")
+    assert np.max(epoch_scores[epoch:]) <= 0.25
+    assert epoch_scores[epoch - 1] > 0.25
+    assert metrics.iteration_finder(epoch_scores, 0.001, "nominal", "always", "below") is None
+
+
+def test_if_nom_first_above():
+    reverse_scores = 1-np.array(epoch_scores)
+    epoch = metrics.iteration_finder(reverse_scores, 0.75, "nominal", "first", "above")
+    assert reverse_scores[epoch] >= 0.75
+    assert reverse_scores[epoch - 1] < 0.75
+    assert metrics.iteration_finder(reverse_scores, 0.999, "nominal", "first", "above") is None
+
+
+def test_if_nom_always_above():
+    reverse_scores = 1 - np.array(epoch_scores)
+    epoch = metrics.iteration_finder(reverse_scores, 0.75, "nominal", "always", "above")
+    assert np.min(reverse_scores[epoch:]) >= 0.75
+    assert reverse_scores[epoch - 1] < 0.75
+    assert metrics.iteration_finder(reverse_scores, 0.999, "nominal", "always", "above") is None
+
+def test_if_ratio_first_below():
+    epoch = metrics.iteration_finder(epoch_scores, 0.5, "ratio", "first", "below")
+    epoch_ratios = np.array(epoch_scores) / epoch_scores[0]
+    assert epoch_ratios[epoch] <= 0.5
+    assert epoch_ratios[epoch - 1] > 0.5
```

Comment:
you should check `np.all(epoch_rates[:epoch] > 0.5)` for 'first' to make it a better test.
Similarly in other tests, check for `epoch_rates[epoch:]` for always

Reply:
Ah... good ideas. Even more thorough.